### PR TITLE
Added `showCheckboxColumn` parameter to DataTable and PaginatedDataTable

### PR DESCRIPTION
**Pull request by bkonyi**
## Description

The `showCheckboxColumn` property will determine whether or not selectable rows in DateTable or PaginatedDataTable will contain a leading CheckBox.

**showCheckboxColumn == true:**
Screen Shot 2019-12-20 at 2 13 38 PM

**showCheckboxColumn == false:**
Screen Shot 2019-12-20 at 2 19 05 PM

## Related Issues

N/A

## Tests

I added the following tests:

*  `DataTable control test - no checkboxes`

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/data_table.dart [hunk 1]
```diff
@@ -408,6 +410,14 @@ class DataTable extends StatelessWidget {
   /// This value defaults to 56.0 to adhere to the Material Design specifications.
   final double columnSpacing;
 
+  /// Whether the widget should display checkboxes for selectable rows.
+  ///
+  /// If true, a [CheckBox] will be placed at the beginning of each row that is
+  /// selectable.
+  ///
+  /// If false, all rows will not display a [CheckBox].
+  final bool showCheckboxColumn;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/data_table.dart]: this should say something about how even if it's true, the column won't be included if there's no row with an `onSelectChanged` callback set.

File: packages/flutter/lib/src/material/paginated_data_table.dart [hunk 2]
```diff
@@ -164,6 +166,14 @@ class PaginatedDataTable extends StatefulWidget {
   /// This value defaults to 56.0 to adhere to the Material Design specifications.
   final double columnSpacing;
 
+  /// Whether the widget should display checkboxes for selectable rows.
+  ///
+  /// If true, a [CheckBox] will be placed at the beginning of each row that is
+  /// selectable.
+  ///
+  /// If false, all rows will not display a [CheckBox].
+  final bool showCheckboxColumn;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/paginated_data_table.dart]: ditto (or better, use doc macros/templates to just insert the text automatically in both places)

File: packages/flutter/test/material/data_table_test.dart [hunk 3]
```diff
@@ -99,6 +99,85 @@ void main() {
...
+      home: Material(child: buildTable(sortColumnIndex: 1, sortAscending: false)),
+    ));
+    await tester.pumpAndSettle(const Duration(milliseconds: 200));
+
+    await tester.tap(find.text('375'));
+
+    expect(log, <String>['cell-tap: 375']);
+    log.clear();
+  });
+
```
Comment (Hixie) [inline packages/flutter/test/material/data_table_test.dart]: should also test that the layout updates correctly if you toggle showCheckboxColumn

**Comment by Hixie** [comment]
approach SGTM

File: packages/flutter/lib/src/material/data_table.dart (hunk 1 above)
Comment (bkonyi) [reply packages/flutter/lib/src/material/data_table.dart]: Done.

File: packages/flutter/lib/src/material/paginated_data_table.dart (hunk 2 above)
Comment (bkonyi) [reply packages/flutter/lib/src/material/paginated_data_table.dart]: TIL about doc templates :+1: . Done!

File: packages/flutter/test/material/data_table_test.dart (hunk 3 above)
Comment (bkonyi) [reply packages/flutter/test/material/data_table_test.dart]: Done.

**Comment by bkonyi** [comment]
Not sure why I'm having some golden tests fail. @Piinks would you happen to know if I'm just holding something wrong?

**Comment by Piinks** [comment]
Looks some new golden files landed and weren't triaged. Should be all good now. I've re-run the failed tasks. :)

**Comment by bkonyi** [comment]
> Looks some new golden files landed and weren't triaged. Should be all good now. I've re-run the failed tasks. :)

My hero! Thanks Kate :-)

File: packages/flutter/test/material/data_table_test.dart [hunk 4]
```diff
@@ -99,6 +99,69 @@ void main() {
...
+            ],
+          );
+        }).toList(),
+      );
+    }
+
+    await tester.pumpWidget(MaterialApp(
+      home: Material(child: buildTable()),
+    ));
+
```
Comment (goderbauer) [inline packages/flutter/test/material/data_table_test.dart]: assert here that `find.byType(Checkbox)` finds nothing?

File: packages/flutter/test/material/data_table_test.dart [hunk 5]
```diff
@@ -99,6 +99,69 @@ void main() {
...
+    expect(log, <String>['row-selected: Cupcake']);
+    log.clear();
+
+    await tester.pumpWidget(MaterialApp(
+      home: Material(child: buildTable(checkboxes: true)),
+    ));
+
+    await tester.pumpAndSettle(const Duration(milliseconds: 200));
+    await tester.tap(find.byType(Checkbox).first);
```
Comment (goderbauer) [inline packages/flutter/test/material/data_table_test.dart]: assert that `find.byType(Checkbox)` finds the number of checkboxes you expect?

**Review by goderbauer — APPROVED** [review]
LGTM modulo minor comments

File: packages/flutter/test/material/data_table_test.dart [hunk 6]
```diff
@@ -99,6 +99,73 @@ void main() {
...
+          );
+        }).toList(),
+      );
+    }
+
+    await tester.pumpWidget(MaterialApp(
+      home: Material(child: buildTable()),
+    ));
+
+    //print(find.byType(Checkbox).allCandidates.length);
```
Comment (goderbauer) [inline packages/flutter/test/material/data_table_test.dart]: Remove the print?

File: packages/flutter/test/material/data_table_test.dart [hunk 7]
```diff
@@ -99,6 +99,73 @@ void main() {
...
+    //expect(find.byType(Checkbox).allCandidates.isEmpty, true);
```
Comment (goderbauer) [inline packages/flutter/test/material/data_table_test.dart]:
```suggestion
    expect(find.byType(Checkbox), findsNothing);
```

File: packages/flutter/test/material/data_table_test.dart [hunk 8]
```diff
@@ -99,6 +99,73 @@ void main() {
...
+    expect(log, <String>['row-selected: Cupcake']);
+    log.clear();
+
+    await tester.pumpWidget(MaterialApp(
+      home: Material(child: buildTable(checkboxes: true)),
+    ));
+
+    await tester.pumpAndSettle(const Duration(milliseconds: 200));
+    final Finder checkboxes = find.byType(Checkbox);
+    expect(checkboxes.evaluate().length, 11);
```
Comment (goderbauer) [inline packages/flutter/test/material/data_table_test.dart]:
```suggestion
    expect(checkboxes, findsNWidgets(11));
```

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**